vpr/src/device/rr_graph_builder.h

```diff
@@ -0,0 +1,45 @@
+#ifndef RR_GRAPH_BUILDER_H
+#define RR_GRAPH_BUILDER_H
+
+#include "rr_graph_storage.h"
+#include "rr_spatial_lookup.h"
+
+/* A data structure allows data modification on a routing resource graph 
+ *
+ * Note that the builder does not own the storage
+ * It serves a virtual protocol for
+ * - node_storage: store the node list 
+ * - node_lookup: store a fast look-up for the nodes
+ *
+ * Note:
+ * - This is the only data structre allowed to modify a routing resource graph
+ *
+ */
+class RRGraphBuilder {
+    /****************
+     * Constructors
+     ****************/
+  public:
+    RRGraphBuilder(t_rr_graph_storage& node_storage,
+                   RRSpatialLookup& node_lookup);
+
+    /****************
+     * Mutators
+     ****************/
+  public:
+    /* Return a writable object for rr_nodes */
+    t_rr_graph_storage& node_storage();
+    /* Return a writable object for update the fast look-up of rr_node */
+    RRSpatialLookup& node_lookup();
+
+    /****************
+     * internal data storage
+     ****************/
+  private:
+    /* node-level storage including edge storages */
+    t_rr_graph_storage& node_storage_;
```

Comment:
Only use references if they are const. If you have a modifiable object, use pointers instead (essentially like you did before).

Using non-const references creates a lot of confusion, in particular if used as instance variables.

Reply:
I fully agree this convention. This rule can improve code readability and help developers to know what they are doing easily. However, as you see in the figure I showed (copied to this PR as well), the builder function will be the owner of the data storages. That is why I temporarily use the reference here. It can avoid a lot of code changes once the refactoring is finished (there is no function get data directly through the ``node_storage`` in ``DeviceContext``). My concern is that if I use pointers, it may cause many codes in client functions or inside the data structures to be changed later. 

Here are my concerns and I am open to discussion.